File: StraussModule/core/Graph.js
export class Graph {

    constructor(nodeCount) {
        this.nodeCount = nodeCount;
        this.adjacency = Array.from({ length: nodeCount }, () => new Set());
        this.edgeCount = 0;
        this.triangleCount = 0;
        this.degrees = new Uint32Array(nodeCount);
    }

    hasEdge(i, j) {
        return this.adjacency[i].has(j);
    }

    addEdge(i, j) {
        if (i === j || this.hasEdge(i, j)) return false;
        const deltaTriangles = this.commonNeighbors(i, j);
        this.adjacency[i].add(j);
        this.adjacency[j].add(i);
        this.edgeCount++;
        this.triangleCount += deltaTriangles;
        this.degrees[i]++;
        this.degrees[j]++;

        return true;
    }

    removeEdge(i, j) {
        if (!this.hasEdge(i, j)) return false;
        const deltaTriangles = this.commonNeighbors(i, j);
        this.adjacency[i].delete(j);
        this.adjacency[j].delete(i);
        this.edgeCount--;
        this.triangleCount -= deltaTriangles;
        this.degrees[i]--;
        this.degrees[j]--;

        return true;
    }

    toggleEdge(i, j) {
        if (this.hasEdge(i, j)) {
            return this.removeEdge(i, j);
        }

        return this.addEdge(i, j);
    }

    degree(i) {
        return this.degrees[i];
    }

    countEdges() {
        return this.edgeCount;
    }

    countTriangles() {
        return this.triangleCount;
    }

    commonNeighbors(i, j) {
        let count = 0;
        const smaller = this.degrees[i] < this.degrees[j] ? this.adjacency[i] : this.adjacency[j];
        const larger = smaller === this.adjacency[i] ? this.adjacency[j] : this.adjacency[i];

        for (const node of smaller) {
            if (larger.has(node)) count++;
        }

        return count;
    }

    getRandomPair() {
        let i = Math.floor(Math.random() * this.nodeCount);
        let j = Math.floor(Math.random() * this.nodeCount);

        while (i === j) {
            j = Math.floor(Math.random() * this.nodeCount);
        }

        return [i, j];
    }
    
    print() {
        for (let i = 0; i < this.nodeCount; i++) {
            console.log(`${i}: ${[...this.adjacency[i]].join(" ")}`);
        }
    }

    printMatrix() {
        for (let i = 0; i < this.nodeCount; i++) {
            console.log(
                Array.from({ length: this.nodeCount }, (_, j) =>
                    this.hasEdge(i, j) ? 1 : 0
                ).join(" ")
            );
        }
    }

    validate() {
        let edges = 0;

        for (let i = 0; i < this.nodeCount; i++) {
            for (const j of this.adjacency[i]) {
                if (!this.adjacency[j].has(i)) {
                    throw new Error("Graph inconsistency");
                }
                if (j > i) edges++;
            }
        }

        if (edges !== this.edgeCount) {
            throw new Error("Wrong edge count");
        }
    }
    getDegreeDistribution(){
        const degrees=[];
        for(let i=0;i<this.nodeCount;i++){
            let degree=0;
            for(let j=0;j<this.nodeCount;j++){
                if(this.hasEdge(i,j)){
                    degree++;
                }
            }
            degrees.push(degree);
        }

        return degrees;
    }
}

File: StraussModule/python_core/Graph.py
import numpy as np
import random


class Graph:

    def __init__(self, nodeCount):
        self.nodeCount = nodeCount
        self.adjacency = np.zeros((nodeCount, nodeCount), dtype=np.uint8)
        self.edgeCount = 0
        self.triangleCount = 0
        self.degrees = np.zeros(nodeCount, dtype=np.uint32)

    def hasEdge(self, i, j):
        return self.adjacency[i, j] == 1

    def addEdge(self, i, j):
        if i == j or self.hasEdge(i, j):
            return False

        deltaTriangles = self.commonNeighbors(i, j)
        self.adjacency[i, j] = 1
        self.adjacency[j, i] = 1
        self.edgeCount += 1
        self.triangleCount += deltaTriangles
        self.degrees[i] += 1
        self.degrees[j] += 1

        return True


    def removeEdge(self, i, j):
        if not self.hasEdge(i, j):
            return False

        deltaTriangles = self.commonNeighbors(i, j)
        self.adjacency[i, j] = 0
        self.adjacency[j, i] = 0
        self.edgeCount -= 1
        self.triangleCount -= deltaTriangles
        self.degrees[i] -= 1
        self.degrees[j] -= 1

        return True


    def toggleEdge(self, i, j):
        if self.hasEdge(i, j):
            return self.removeEdge(i, j)

        return self.addEdge(i, j)


    def degree(self, i):
        return int(self.degrees[i])


    def countEdges(self):
        return self.edgeCount


    def countTriangles(self):
        return self.triangleCount


    def commonNeighbors(self, i, j):
        count = 0
        for node in range(self.nodeCount):
            if self.adjacency[i, node] and self.adjacency[j, node]:
                count += 1

        return count


    def getRandomPair(self):
        i = random.randrange(self.nodeCount)
        j = random.randrange(self.nodeCount)

        while i == j:
            j = random.randrange(self.nodeCount)

        return i, j


    def getEdges(self):
        edges = []

        for i in range(self.nodeCount):
            for j in range(i + 1, self.nodeCount):
                if self.adjacency[i, j]:
                    edges.append([i, j])

        return edges


    def print(self):
        for i in range(self.nodeCount):
            neighbors = np.where(self.adjacency[i] == 1)[0]
            print(f"{i}: {' '.join(map(str, neighbors))}")

    def printMatrix(self):
        for i in range(self.nodeCount):
            print(" ".join(str(int(value)) for value in self.adjacency[i]))


    def validate(self):
        edges = 0

        for i in range(self.nodeCount):
            for j in range(i + 1, self.nodeCount):

                if self.adjacency[i, j] != self.adjacency[j, i]:
                    raise RuntimeError("Graph inconsistency")

                if self.adjacency[i, j]:
                    edges += 1

        if edges != self.edgeCount:
            raise RuntimeError("Wrong edge count")


    def getDegreeDistribution(self):
        degrees = []

        for i in range(self.nodeCount):
            degree = 0

            for j in range(self.nodeCount):
                if self.hasEdge(i, j):
                    degree += 1

            degrees.append(degree)

        return degrees
